### PR TITLE
[2201.2.x] Disallow desugar query expression twice for method invocation

### LINKED ISSUE #34225: Compiler throws AssertionError when compiling valid code
**Description:**

When compiling [this valid code](https://github.com/spencerwi/AdventOfCode2021/blob/250e1b2c9c46ee45da01dbdbdf74c500a1d0bbda/day8/main.bal), the compiler crashes with the following error message:

```
Compiling source
        vscode/day8:0.1.0
ballerina: Oh no, something really went wrong. Bad. Sad.

We appreciate it if you can report the code that broke Ballerina in
https://github.com/ballerina-platform/ballerina-lang/issues with the
log you get below and your sample code.

We thank you for helping make us better.

[2021-12-08 15:28:16,578] SEVERE {b7a.log.crash} - null 
java.lang.AssertionError
        at org.wso2.ballerinalang.compiler.tree.BLangNodeVisitor.visit(BLangNodeVisitor.java:630)
        at org.wso2.ballerinalang.compiler.tree.expressions.BLangFieldBasedAccess.accept(BLangFieldBasedAccess.java:65)
        at org.wso2.ballerinalang.compiler.bir.BIRGen.visit(BIRGen.java:1675)
        at org.wso2.ballerinalang.compiler.tree.expressions.BLangTypeConversionExpr.accept(BLangTypeConversionExpr.java:72)
        at org.wso2.ballerinalang.compiler.bir.BIRGen.visit(BIRGen.java:1179)
        at org.wso2.ballerinalang.compiler.tree.statements.BLangSimpleVariableDef.accept(BLangSimpleVariableDef.java:46)
        at org.wso2.ballerinalang.compiler.bir.BIRGen.visit(BIRGen.java:741)
        at org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody.accept(BLangBlockFunctionBody.java:57)
        at org.wso2.ballerinalang.compiler.bir.BIRGen.visit(BIRGen.java:697)
        at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:73)
        at org.wso2.ballerinalang.compiler.bir.BIRGen.lambda$visit$5(BIRGen.java:396)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
        at org.wso2.ballerinalang.compiler.bir.BIRGen.visit(BIRGen.java:396)
        at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:163)
        at org.wso2.ballerinalang.compiler.bir.BIRGen.genBIR(BIRGen.java:266)
        at io.ballerina.projects.internal.CompilerPhaseRunner.birGen(CompilerPhaseRunner.java:220)
        at io.ballerina.projects.internal.CompilerPhaseRunner.performBirGenPhases(CompilerPhaseRunner.java:157)
        at io.ballerina.projects.ModuleContext.generateCodeInternal(ModuleContext.java:416)
        at io.ballerina.projects.ModuleCompilationState$4.generatePlatformSpecificCode(ModuleCompilationState.java:132)
        at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:353)
        at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:159)
        at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:137)
        at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:113)
        at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
        at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:182)
        at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:112)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:80)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:191)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at io.ballerina.cli.launcher.Main.main(Main.java:51)
 
ERROR [day8:(1:1,1:1)] Compilation failed due to an unhandled exception
error: compilation contains errors
```

**Steps to reproduce:**

1. Launch a new Codespace from [this repo at this commit](https://github.com/spencerwi/AdventOfCode2021/tree/250e1b2c9c46ee45da01dbdbdf74c500a1d0bbda/) (there's already devcontainer configuration in the repo)
2. In that codespace, run `cd day8 && bal run`
3. Observe the error message presented above.

**Affected Versions:**
Swan Lake Beta 4 (possibly also Swan Lake Beta 5; there's no Beta 5 devcontainer image available yet though)

**OS, DB, other environment details and versions:**
Using the Swan Lake Beta 4 devcontainer docker image.

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by prakanth97**
## Purpose

Fixes #34225

## Samples
After this change this code will not give the compilation error.
```
string[] words = ["a", "b", "c"];
var len = (from string str in words
    select str).length();
```

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 1]
```diff
@@ -6587,6 +6587,11 @@ private void rewriteInvocation(BLangInvocation invocation, boolean async) {
         reorderArguments(invocation);
 
         rewriteExprs(invocation.requiredArgs);
+        if (invocation.langLibInvocation && !invocation.requiredArgs.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (prakanth97) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:6590]: When we have a query-return(or a variable) invokes Langlib function like this,
```
string[] words = ["a", "b", "c"];
var len = (from string str in words
    select str).length();
```
According to the previous logic it desugars twice (`expr` and first `required arg` same) and it causes issue #34225. Here we check if it is Langlib invocation and `required args` is not empty, then we use the already desugared output.

I will put a comment to explain this

[the remark this answers is not in the document]
Comment (prakanth97) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:6593]: Updated 96bf1ce3588d770ae39e3fbd51858302c29807d2

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 2]
```diff
@@ -6587,6 +6587,14 @@ private void rewriteInvocation(BLangInvocation invocation, boolean async) {
         reorderArguments(invocation);
 
         rewriteExprs(invocation.requiredArgs);
+
+        // Disallow desugar expr again for langlib method invocation because
+        // it is the only place where expr and first required args same.
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:6592]:
```suggestion
        // Disallow desugaring the same exprression twice. 
        // For langlib invocations, expr is duplicated as the first required arg
```

**Review by pcnfernando — APPROVED** [review]
Add 0456b6a3f8b29f611f0c186429e9760a4e82d14f to master as well

**Pull request merged**